burn-dataset/src/transform/sampler.rs

```diff
@@ -1,38 +1,77 @@
-use std::{marker::PhantomData, sync::Mutex};
+use rand::{distributions::Uniform, rngs::StdRng, seq::IteratorRandom, Rng, SeedableRng};
+use std::{marker::PhantomData, ops::DerefMut, sync::Mutex};
 
-/// Sample items from a dataset with replacement.
+/// Sample items from a dataset.
 ///
-/// This is an efficient way of modeling a dataset as a probability distribution of a fixed size.
+/// This is an convenient way of modeling a dataset as a probability distribution of a fixed size.
+/// You have multiple options to instantiate the dataset sampler.
+///
+/// * With replacement: This is the most efficient way of using the sampler because no state is
+///   required to keep indices that have been selected.
+///
+/// * Without replacement: This has a similar effect to using a
+///   [shuffled dataset](crate::transform::ShuffledDataset), but with more flexibility since you can
+///   set the dataset to an arbitrary size.
 pub struct SamplerDataset<D, I> {
     dataset: D,
     size: usize,
-    rng: Mutex<StdRng>,
+    state: Mutex<SamplerState>,
     input: PhantomData<I>,
 }
 
+enum SamplerState {
+    WithReplacement(StdRng),
+    WithoutReplacement(StdRng, Vec<usize>),
+}
+
 impl<D, I> SamplerDataset<D, I>
 where
     D: Dataset<I>,
     I: Send + Sync,
 {
-    /// Creates a new sampler dataset.
+    /// Creates a new sampler dataset with replacement.
     pub fn new(dataset: D, size: usize) -> Self {
-        let rng = Mutex::new(StdRng::from_entropy());
+        Self {
+            dataset,
+            size,
+            state: Mutex::new(SamplerState::WithReplacement(StdRng::from_entropy())),
+            input: PhantomData,
+        }
+    }
+
+    /// Creates a new sampler dataset with replacement.
+    pub fn with_replacement(dataset: D, size: usize) -> Self {
+        Self::new(dataset, size)
+    }
 
+    /// Creates a new sampler dataset without replacement.
+    pub fn without_replacement(dataset: D, size: usize) -> Self {
         Self {
             dataset,
             size,
-            rng,
+            state: Mutex::new(SamplerState::WithoutReplacement(
+                StdRng::from_entropy(),
+                Vec::new(),
+            )),
             input: PhantomData,
         }
     }
 
-    /// Generates random index using uniform distribution (0, dataset.len()).
     fn index(&self) -> usize {
-        let mut rng = self.rng.lock().unwrap();
-        rng.sample(Uniform::new(0, self.dataset.len()))
+        let mut state = self.state.lock().unwrap();
+
+        match state.deref_mut() {
+            SamplerState::WithReplacement(rng) => rng.sample(Uniform::new(0, self.dataset.len())),
+            SamplerState::WithoutReplacement(rng, indices) => {
+                if indices.is_empty() {
+                    // Refill the state.
+                    *indices = (0..self.dataset.len()).choose_multiple(rng, self.dataset.len());
+                }
+
+                indices.pop().expect("State is never empty")
```

Comment:
Again, we should document the fact that WithoutReplacement actually starts over when all indices have been picked once. 
Also, "State is never empty" is a bit cryptic to me